operator/pkg/listener/watcher_events.go

```diff
@@ -0,0 +1,147 @@
+				"cluster-name", clusterName,
+			).Error(err, "could not trigger the reconciliation loop")
+			return
+		}
+
+		h.logger.WithValues(
+			"component", componentName,
+			"cluster-name", clusterName,
+		).Info("successfully triggered component reconciliation loop")
+	}
+}
+
+func (h *WatcherEventsHandler) updateComponentCRLabels(ctx context.Context, componentCR *unstructured.Unstructured, kymaCR *v1alpha1.Kyma, q workqueue.RateLimitingInterface) error {
+	componentLabels := componentCR.GetLabels()
+	labelValue := strings.ReplaceAll(time.Now().UTC().Format(time.RFC3339), ":", ".")
+	componentLabels[labels.ListenerLastUpdated] = labelValue
+	componentCR.SetLabels(componentLabels)
+	err := h.Update(ctx, componentCR, &client.UpdateOptions{})
+	if client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("could not update component CR: %s", err)
+	}
+	if errors.IsNotFound(err) {
+		componentKey := client.ObjectKeyFromObject(componentCR)
+		err = h.handleComponentNotFoundError(ctx, kymaCR, componentKey, q)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *WatcherEventsHandler) getComponentCR(ctx context.Context, kymaCR *v1alpha1.Kyma, componentName, clusterName string, q workqueue.RateLimitingInterface) (*unstructured.Unstructured, error) {
+	templates, err := release.GetTemplates(ctx, h.Client, kymaCR)
+	if err != nil {
+		return nil, err
+	}
+
+	componentCR, err := util.GetUnstructuredComponentFromTemplate(templates, componentName, kymaCR)
+	if err != nil {
+		return nil, err
+	}
+
+	componentKey := client.ObjectKeyFromObject(componentCR)
+	err = h.Get(ctx, componentKey, componentCR)
+	if client.IgnoreNotFound(err) != nil {
+		return nil, fmt.Errorf("could not get component CR: %s", err)
+	}
+	if errors.IsNotFound(err) {
+		err = h.handleComponentNotFoundError(ctx, kymaCR, componentKey, q)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return componentCR, nil
+}
+
+func (h *WatcherEventsHandler) getKymaCR(ctx context.Context, clusterName string) (*v1alpha1.Kyma, error) {
+	kymaCRsForCluster := &v1alpha1.KymaList{}
+	err := h.List(ctx, kymaCRsForCluster, client.MatchingLabels{
+		labels.ClusterName: clusterName,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not get Kyma CR for cluster: %s", err)
+	}
+	if len(kymaCRsForCluster.Items) == 0 {
+		return nil, fmt.Errorf("kyma CR for cluster not found")
+	}
+	if len(kymaCRsForCluster.Items) > 1 {
+		return nil, fmt.Errorf("found more than one Kyma CR for cluster")
+	}
+	return &kymaCRsForCluster.Items[0], nil
+}
+
+func (h *WatcherEventsHandler) handleComponentNotFoundError(ctx context.Context, kymaCR *v1alpha1.Kyma, componentKey client.ObjectKey, q workqueue.RateLimitingInterface) error {
+	latestKymaCR := &v1alpha1.Kyma{}
+	if err := h.Get(ctx, client.ObjectKeyFromObject(kymaCR), latestKymaCR); err != nil {
+		return fmt.Errorf("could not get updated Kyma CR for component: %s", err)
+	}
+
+	if kymaCR.Status.State != v1alpha1.KymaStateProcessing {
```

Comment:
This is a big assumption, You won't know what the state of the Kyma will be in when it gets taken out of the  for processing. Why not requeue it every time you receive an event? Is this a load issue?

Reply:
As mentioned above. The new design approach resolves the above concerns!